### ISSUE TITLE
STMAP exporter

### ISSUE THREAD

**Issue by bhack** [opencv/opencv#24439]
### Describe the feature and motivation

STmaps[1][2] are quite popular in many pipelines but it seems hard to replicate what it is expected to be processed by other software like Natron, Blender, Nuke, etc.
It seems that exporting it with `initUndistortRectifyMap` it will not have the same effect in these software as OpenCV `initUndistortRectifyMap` + `remap` also using the same interpolation and datatype precision. 

It is going always to have substantial visual difference. 

[1]https://benmcewan.com/blog/2020/02/02/demystifying-st-maps/
[²]https://natron.readthedocs.io/en/rb-2.5/plugins/net.sf.openfx.STMap.html

### Additional context

It could be partially connected to https://github.com/opencv/opencv/issues/22120 as we need a direct mapping and not the inverse one expected by `remap`

**Comment by bhack** [comment]
I am going to reopen this as I've seen that we have introduced a new API ` initInvertedRectificationMap` to invert `initUndistortRectifyMap` in `4.5.3`.

But I suppose that we could still have a problem:

1. Get the required/estimated size for the undistorted image with undistortpoints (min and max):
    `cv::undistortPoints(
      distortedPoints, undistortedPoints, cameraMatrix, distCoeffs,
      cv::noArray(),
      cameraMatrix);`
 2.   Get the Optimal new camera matrix `newCameraMatrix = cv::getOptimalNewCameraMatrix(
        cameraMatrix, distCoeffs, cv::Size(originalWidth, originalHeight), 1,
        estimatedSize, 0, true);`
3.    Prepare the undistortion map. We are using the inverse here as the API prepare it for `remap` that use the inverse mapping:
        `cv::initInverseRectificationMap(cameraMatrix, distCoeffs, cv::Mat(),
                                    newCameraMatrix, estimatedSize, CV_32FC2,
                                    map, cv::noArray());`
4. Here I think we could have a problem when we are preparing the distortion map. I suppose we need to get the optimal camera matrix to apply back the distortion and to reproduce the original source image at the original size. Bur `getOptimalNewCameraMatrix`  uses `icvGetRectangles` that internally has only `cvUndistortPoints`. 

So how we are going to take the new optimal camera matrix to apply the distortion and go back (up to the interpolation artifacts) to the original image?

/cc @vpisarev  Any feedback on this?

**Comment by IanMaquignaz** [comment]
The function initInvertedRectificationMap is semantically the opposite of initUndistortRectifyMap, but I'm not sure it applies to your context. When I wrote the function, I was designing ProCam (Projector-Camera) pairs and needed a method of performing the equivalent of initUndistortRectifyMap on a projector (which is essentially an inverse camera). Therefore all the operations are inversed but also applied in the inverse order to model the image progression from a projector's 3LCD emitter, through the optics, and then finally appear on an incident surface (e.g. a wall) while appearing 'undistorted'.

Be very careful and make sure this is what you actually want.

**Comment by bhack** [comment]
I was just trying to produce the maps to undistort the image and then to distort it again to produce back the original image.

**Comment by IanMaquignaz** [comment]
I believe my original test case was proving no error in transformation between `cv::initUndistortRectifyMap`  <-> `cv::initInvertedRectificationMap`. A dependence on `cv::initUndistortRectifyMap` was not desirable for the test case, but [you can find the test code here](https://github.com/opencv/opencv/blob/617d7ff575200c0a647cc615b86003f10b64587b/modules/calib3d/test/test_undistort.cpp#L822). 

My suggestion to you is to decompose the function into individual transformations and incrementally build the transformation pipeline you need.

**Comment by bhack** [comment]
> I believe my original test case was proving no error in transformation between cv::initUndistortRectifyMap <-> cv::initInvertedRectificationMap

But have you also tested how to keep pixels that were going outside the original resolution?
Cause this requires to calculate the undistorted camera optimal size and a new optimal camera matrix for the `newcameraMatrix` at `cv::initUndistortRectifyMap`.  
If not, how are you going to recover all the pixels for `cv::initInvertedRectificationMap`?

**Comment by IanMaquignaz** [comment]
My scope was Procam-System. If the pixels were distorted  beyond the projector or camera image size, they were lost. That being said, my camera resolution was 3x that of the projector and I would have proven that case. 

Have you created a test case yet?

**Comment by bhack** [comment]
Yes but I cannot exactly go back to the same original image with the maps produced by these functions.

If you want to go from the original distorted source to the undistort one and then back to distort (after an edit) with these two maps you don't need lose any pixel in the undistort map.

That is the role of `alpha` Param in:

https://docs.opencv.org/4.8.0/d9/d0c/group__calib3d.html#ga7a6c4e032c97f03ba747966e6ad862b1

But is this function really valid for your inverse?

**Comment by bhack** [comment]
So basically when you use the same `cameramatrix` and the same original image size it was almost giving "almost" back the original distorted image:
- `initRectificatyUndistortMap(cameraMatrix, distCoeffs, cv::Mat(), cameraMatrix, originalImgSize, CV_32FC2, uvMap, dummyMap)` 
- the same for `initInveseRectificationMap` 

The problem is when you instead introduce `cv::getOptimalNewCameraMatrix` to control the alpha/scale parameters.
Are we sure that we are correctly handling in`initInveseRectificationMap`? 
Do you we had test converge for this case?
